### PR TITLE
Update quickplot to catch empty lists

### LINKED ISSUE #3172: Raise error if empty list passed to `QuickPlot`
If you pass any empty list to `QuickPlot` you get the error:
```
  File "/Users/robertwtimms/Documents/PyBaMM/test.py", line 3, in <module>
    pybamm.QuickPlot([])
  File "/Users/robertwtimms/Documents/PyBaMM/pybamm/plotting/quick_plot.py", line 146, in __init__
    output_variables = models[0].default_quick_plot_variables
IndexError: list index out of range
```

It would be nicer to catch this with a more informative error earlier so you can check if you've made a silly mistake like I did where I wasn't appending solutions to a list.

Code to reproduce:
```python3
import pybamm

pybamm.QuickPlot([])
```

### PR DESCRIPTION
**Pull request by kratman**
# Description

I added some error handling for QuickPlot when an empty list is passed as the first argument.

Changes:
- Added a check that raises an error when an empty list is passed: ```QuickPlot([])```
- Added a new test to confirm the check functions correctly.
- Minor refactoring of QuickPlot to create new functions for the input pre-processing.

Fixes #3172

## Type of change

Minor bug fix for catching empty input lists.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: tests/unit/test_plotting/test_quick_plot.py [hunk 1]
```diff
@@ -464,9 +464,13 @@ def test_plot_2plus1D_spm(self):
         pybamm.close_plots()
 
     def test_failure(self):
```
Comment (brosaplanella) [inline tests/unit/test_plotting/test_quick_plot.py:466]: Can you call it something a bit more explicit (e.g. `test_wrong_type` or something like this)?

**Review by brosaplanella — COMMENTED** [review]
Looks good, just a couple of things before approving. Can you address the small comment on the test and add an entry to the CHANGELOG?

File: tests/unit/test_plotting/test_quick_plot.py (hunk 1 above)
Comment (kratman) [reply tests/unit/test_plotting/test_quick_plot.py:466]: Done

**Review by brosaplanella — APPROVED** [review]
Can you commit the suggested change? Happy to merge afterwards

**Comment by brosaplanella** [comment]
@all-contributors please add @kratman for tests

**Pull request merged**